tools/hlstool/hlstool.cpp

```diff
@@ -0,0 +1,451 @@
+                      cl::init(false), cl::Hidden);
+
+static cl::opt<bool>
+    verbosePassExecutions("verbose-pass-executions",
+                          cl::desc("Log executions of toplevel module passes"),
+                          cl::init(false));
+
+static cl::opt<bool>
+    verifyPasses("verify-each",
+                 cl::desc("Run the verifier after each transformation pass"),
+                 cl::init(true));
+
+static cl::opt<bool>
+    allowUnregisteredDialects("allow-unregistered-dialects",
+                              cl::desc("Allow unknown dialects in the input"),
+                              cl::init(false), cl::Hidden);
+
+enum HLSFlow { HLSFlowDynamic };
+
+static cl::opt<HLSFlow> hlsFlow(
+    cl::desc("HLS flow"),
+    cl::values(clEnumValN(HLSFlowDynamic, "dynamic", "Dynamically scheduled")));
+
+enum OutputFormatKind { OutputIR, OutputVerilog };
+
+static cl::opt<int>
+    irInputLevel("ir-input-level",
+                 cl::desc("Level at which to input IR at. It is flow-defined "
+                          "which value corersponds to which IR level."),
+                 cl::init(-1));
+
+static cl::opt<int>
+    irOutputLevel("ir-output-level",
+                  cl::desc("Level at which to output IR at. It is flow-defined "
+                           "which value corersponds to which IR level."),
+                  cl::init(-1));
+
+static cl::opt<OutputFormatKind> outputFormat(
+    cl::desc("Specify output format:"),
+    cl::values(clEnumValN(OutputIR, "ir", "Emit post-HLS IR"),
+               clEnumValN(OutputVerilog, "verilog", "Emit Verilog")),
+    cl::init(OutputVerilog));
+
+// --------------------------------------------------------------------------
+// (Configurable) pass pipelines
+// --------------------------------------------------------------------------
+
+/// Create a simple canonicalizer pass.
+static std::unique_ptr<Pass> createSimpleCanonicalizerPass() {
+  mlir::GreedyRewriteConfig config;
+  config.useTopDownTraversal = true;
+  config.enableRegionSimplification = false;
+  return mlir::createCanonicalizerPass(config);
+}
+
+static void loadDHLSPipeline(OpPassManager &pm) {
+  // Software lowering
+  pm.addPass(mlir::createLowerAffinePass());
+  pm.addPass(mlir::createConvertSCFToCFPass());
+  pm.addPass(circt::createFlattenMemRefPass());
+
+  // DHLS conversion
+  pm.addPass(circt::createStandardToHandshakePass());
+}
+
+static void loadHandshakeTransformsPipeline(OpPassManager &pm) {
+  pm.nest<handshake::FuncOp>().addPass(createSimpleCanonicalizerPass());
+  pm.nest<handshake::FuncOp>().addPass(
+      handshake::createHandshakeMaterializeForksSinksPass());
+  pm.nest<handshake::FuncOp>().addPass(createSimpleCanonicalizerPass());
+  // Todo: arguments
+  pm.nest<handshake::FuncOp>().addPass(
+      handshake::createHandshakeInsertBuffersPass());
+  pm.nest<handshake::FuncOp>().addPass(createSimpleCanonicalizerPass());
+}
+
+static void loadFIRRTLLoweringPipeline(OpPassManager &pm) {
```

Comment:
This is great, but it will inevitably drift from the "golden" pipeline in firtool. Not something to worry about right now, but it would be nice to expose this API somewhere in the FIRRTL dialect, so hlstool and firtool can share the same pipeline. Or, we stop using FIRRTL for HLS 😉

Reply:
The inclusion of this will hopefully have a **very** short lifetime, assuming #3851 gains traction.